crates/storage/provider/src/writer/mod.rs

```diff
@@ -188,23 +190,25 @@ where
         //  * trie updates (cannot naively extend, need helper)
         //  * indices (already done basically)
         // Insert the blocks
-        for block in blocks {
-            let sealed_block =
-                block.block().clone().try_with_senders_unchecked(block.senders().clone()).unwrap();
+        for ExecutedBlock { block, senders, execution_output, hashed_state, trie } in blocks {
+            let sealed_block = Arc::unwrap_or_clone(block)
+                .try_with_senders_unchecked(Arc::unwrap_or_clone(senders))
+                .unwrap();
             self.database().insert_block(sealed_block, StorageLocation::Both)?;
 
             // Write state and changesets to the database.
             // Must be written after blocks because of the receipt lookup.
-            let execution_outcome = block.execution_outcome().clone();
-            state_writer.write_to_storage(execution_outcome, OriginalValuesKnown::No)?;
+            state_writer.write_to_storage(
+                Arc::unwrap_or_clone(execution_output),
+                OriginalValuesKnown::No,
+            )?;
 
             // insert hashes and intermediate merkle nodes
-            {
-                let trie_updates = block.trie_updates().clone();
-                let hashed_state = block.hashed_state();
-                self.database().write_hashed_state(&hashed_state.clone().into_sorted())?;
-                self.database().write_trie_updates(&trie_updates)?;
-            }
+            // TODO: These writes should take ownership instead of cloning through
+            // the reference just to (likely) drop the original afterwards.
```

Comment:
unclear what this todo means, what is todo?

Reply:
I meant `write_hashed_state` and `write_trie_updates` should take the full data instead of a reference just to clone it within. That only makes sense if we want to persist the data afterwards, then we only need to clone the values instead of the whole structure. However, in this case, we don't need to persist the input data so passing the whole thing in would avoid any cloning. Should I elaborate or just remove the comment for now?

Reply:
Removed the comment. Will just do it in upcoming PRs that give better context anyway.